### PR TITLE
[CINFRA-810] Migrate replicated log tests

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Introduce a new testing framework for replicated logs, and use it to migrate the following test suites, which are currently disabled:

-  ReplicatedLog/AppendEntriesBatchTest.cpp
-  ReplicatedLog/LeaderAppendEntriesTest.cpp
-  ReplicatedLog/MultiTermTest.cpp
-  ReplicatedLog/RewriteLogTest.cpp

The following files are still disabled, and to be migrated in a later PR:

- ReplicatedLog/ConcurrencyTests.cpp
- ReplicatedLog/SimpleInsertTests.cpp
- ReplicatedLog/UpdateParticipantsFlags.cpp
- ReplicatedLog/WaitForSyncTest.cpp

### CODE REVIEW

**Review by apetenchea — CHANGES_REQUESTED** [review]
Apart from my comment regarding `std::remove_if`, the rest are more or less nitpicks.

File: tests/Replication2/CMakeLists.txt [hunk 1]
```diff
@@ -133,6 +145,8 @@ target_link_libraries(arango_tests_replication2_pure PUBLIC
 
 target_include_directories(arango_tests_replication2_pure PRIVATE "${PROJECT_SOURCE_DIR}/tests")
 
+# This binary currently can't be linked (and couldn't be for a while). We should
+# decide whether to fix it or remove it. [tobias, 2023-08-10]
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/CMakeLists.txt:149]: Opened a ticket: https://arangodb.atlassian.net/browse/CINFRA-850

File: tests/Replication2/Helper/ReplicatedLogTestSetup.h [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+      ParticipantWithFakes const& other) const noexcept -> bool {
+    // suffices for now, we could conceivably also compare logId +
+    // serverInstance instead
+    return this == &other;
+  }
+
+  [[nodiscard]] auto updateConfig(LogConfig const* conf)
+      -> futures::Future<futures::Unit>;
+  [[nodiscard]] auto getAsLeader() {
+    auto const leader = std::dynamic_pointer_cast<replicated_log::ILogLeader>(
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/Helper/ReplicatedLogTestSetup.h]: I don't think it's moved in the return, there should be a copy elision / RVO, which I think should be unaffected by the `const`. I've applied your suggestion anyway.

File: tests/Replication2/Helper/ReplicatedLogTestSetup.h [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+  auto addNewTerm(std::optional<ParticipantId> leader,
+                  std::vector<ParticipantId> follower,
+                  ConfigArguments configArguments) -> LogConfig* {
+    auto const term = configArguments.term;
+    TRI_ASSERT(!terms.contains(term));
+
+    auto& cnt = logs.at(*leader);
+    auto leaderContainer =
+        leader ? std::optional<std::reference_wrapper<LogContainer>>{cnt}
+               : std::nullopt;
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/Helper/ReplicatedLogTestSetup.h]: Right, thanks for spotting.

File: tests/Replication2/Helper/ReplicatedLogTestSetup.h [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+    std::optional<std::uint64_t> setWriteConcern;
+    std::optional<bool> setWaitForSync;
+
+    auto updateConfig(LogConfig config) const -> LogConfig {
+      for (auto const& toRemove : removeParticipants) {
+        auto const shouldBeRemoved = [&](auto other) {
+          return toRemove.get() == other.get();
+        };
+        std::remove_if(config.followers.begin(), config.followers.end(),
+                       shouldBeRemoved);
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/Helper/ReplicatedLogTestSetup.h]: :see_no_evil: Of course. Thanks for noticing!

File: tests/Replication2/ReplicatedLog/MultiTermTest.cpp [hunk 5]
```diff
@@ -241,69 +231,70 @@ struct FollowerProxy : AbstractFollower {
...
-        followerLog->becomeFollower("newFollower", LogTerm{2}, "newLeader");
+    EXPECT_TRUE(
+        followerLogContainer->delayedLogFollower->hasPendingAppendEntries());
+    // TODO this swapping here should be replaced by something in the test
+    //      framework (i.e. LogContainer, LogConfig etc.)
+    auto oldFollower = DelayedLogFollower(followerLogContainer->serverId());
+    auto oldScheduler = DelayedScheduler();
+    followerLogContainer->delayedLogFollower->swapFollowerAndQueueWith(
+        oldFollower);
+    std::swap(oldScheduler, *followerLogContainer->logScheduler);
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/ReplicatedLog/MultiTermTest.cpp]: Right, that happened due to incremental changes regarding this. Thanks, it's cleaner now.

**Review by apetenchea — APPROVED** [review]
LGTM

File: tests/Replication2/Mocks/DelayedLogFollower.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      std::shared_ptr<replicated_log::ILogFollower> follower);
+
+  void swapFollowerAndQueueWith(DelayedLogFollower& other);
+
+  auto appendEntries(replicated_log::AppendEntriesRequest req)
+      -> arangodb::futures::Future<
+          replicated_log::AppendEntriesResult> override;
+
+  // TODO rename to sth like deliver-delayed-append-entries-requests or so
+  auto runAsyncAppendEntries() -> std::size_t;
```
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Replication2/Mocks/DelayedLogFollower.h:63]: Not necessarily.

**Comment by goedderz** [comment]
> LGTM
> 
> I am wondering if we can work around this concept issue in IHasScheduler though...

I haven't found a way in the time I invested - and I think waiting for a recent enough compiler is fine, this will just solve itself.

**Pull request merged**